alternator/executor.hh

```diff
@@ -70,6 +70,42 @@ public:
     std::string to_json() const override;
 };
 
+namespace parsed {
+class path;
+};
+
+// A hierarchy_filter is used to specify only a part of the hierarchy that we
+// wish to extract from a JSON object. For example when a ProjectionExpression
+// requires that we only return certain pieces of a top-level attribute.
+struct hierarchy_filter {
+    // Either "members" or "indexes" may be non-empty, but not both. This is
+    // because DynamoDB does not allow both a.b and a[1] to be requested at
+    // the same time. Those are considered to be "conflicting" document paths.
+    // We could allow both, if we wanted to diverge from DynamoDB.
+    // We need the extra unique_ptr<> here because libstdc++ unordered_map
+    // doesn't work with incomplete types :-(
+    std::unordered_map<std::string, std::unique_ptr<hierarchy_filter>> members;
+    std::unordered_map<unsigned, std::unique_ptr<hierarchy_filter>> indexes;
+    // Add a path to the filter. Throws a validation error if the path
+    // "overlaps" with one already in the filter (one is a sub-path
+    // of the other) or "conflicts" with it (both a member and index is
+    // requested).
+    void add(const parsed::path&);
+    // Run the filter on a given JSON object, dropping its parts which
+    // weren't asked to be kept. Modifies the given JSON object, or
+    // returns false to signify the entire object should be dropped.
+    // Important note: The JSON object is assumed to be using the DynamoDB
+    // conventions, i.e., it is really a map whose key has a type string,
+    // and the value is the real object.
+    bool filter(rjson::value&);
+};
+
+// We use lw_shared_ptr<hierarchy_filter>  allows us to efficiently represent
+// with a nullptr the common case where there is no hierarchy filter, i.e.,
+// a top-level attribute is requested. We use lw_shared_ptr and not unique_ptr
+// to allow copying attrs_to_get.
+using attrs_to_get = std::unordered_map<std::string, lw_shared_ptr<hierarchy_filter>>;
```

Comment:
Just curious - do we actually need to move `attrs_to_get` somewhere in the code? Maybe the copy can simply be elided?

Reply:
You're right. It was hard to debug where these copies are coming from - I get 100 lines of errors and no clear indication of where the copy is actually done. I was able to find just one place in describe_items_visitor()'s constructor. But there is no real reason why it must copy. I'll try to avoid this copy, and if I can I'll go back to using unique_ptr.

Reply:
Actually I was wrong - that constructor doesn't copy - just keeps a pointer. I now suspect the "table_requests" object, but I still can't find exactly where the copy is made. I'm not sure it's worth pursuing this, unless you think it's important.

Reply:
It's not, it's fine to just fix it in a follow-up one day